File: solutions/algorithms/algorithm_bubble_sort.m
#import <Foundation/Foundation.h>

/**
 * Create the array of random numbers
 */
NSMutableArray* createArray(int len, int min, int max) {
  NSMutableArray *arr = [NSMutableArray arrayWithCapacity:len];
  int diff = max - min;

  for (int i = 0; i < len; i++) {
    NSNumber *number = [NSNumber numberWithInt:min + arc4random() % diff];
    [arr insertObject:number atIndex:i];
  }
  return arr;
}

/**
 * Copy and sort the array using native methods
 */
NSMutableArray* getSortedArray(NSArray* array) {
  NSMutableArray *sortedArray = [NSMutableArray arrayWithArray:array];
  NSSortDescriptor *sorter = [NSSortDescriptor sortDescriptorWithKey:@"self" ascending:YES];
  [sortedArray sortUsingDescriptors:[NSArray arrayWithObject:sorter]];
  return sortedArray;
}

/**
 * Simple bubble sort
 */
NSMutableArray* bubbleSort(NSArray* array) {
  NSMutableArray *sortedArray = [NSMutableArray arrayWithArray:array];

  BOOL swapped;
  int len = [sortedArray count];
  do {
    swapped = NO;

    for (int i = 0; i < len - 1; i++) {
      NSNumber* num1 = [sortedArray objectAtIndex:i];
      NSNumber* num2 = [sortedArray objectAtIndex:i+1];
      if ([num2 intValue] < [num1 intValue]) {
        swapped = YES;
        [sortedArray replaceObjectAtIndex:i withObject:num2];
        [sortedArray replaceObjectAtIndex:i+1 withObject:num1];
      }
    }
  } while (swapped);

  return sortedArray;
}

/**
 * Compare the arrays and make sure all values match
 */
BOOL compareArrays(NSMutableArray* array1, NSMutableArray *array2) {
  int len1 = [array1 count];
  int len2 = [array2 count];
  if (len1 != len2) {
    NSLog(@"The array lengths do not match");
    return NO;
  }

  NSNumber *num1;
  NSNumber *num2;
  for (int i = 0; i < len1; i++) {
    num1 = [array1 objectAtIndex:i];
    num2 = [array2 objectAtIndex:i];
    if ([num1 intValue] != [num2 intValue]) {
      NSLog(@"The array items do not match at index %d", i);
      return NO;
    }
  }

  return YES;
}

/**
 * Run the thing
 */
int main (int argc, const char * argv[]) {
  NSMutableArray *arr = createArray(1000, 0, 10000);
  NSMutableArray *verificationArr = getSortedArray(arr);
  NSMutableArray *sortedArr = bubbleSort(arr);

  BOOL sortWorked = compareArrays(sortedArr, verificationArr);
  if (sortWorked) {
    NSLog(@"The sort worked!");
  } else {
    NSLog(@"THE SORT FAILED");
  }

  return 0;
}

File: solutions/algorithms/algorithm_bubble_sort.go
package main

import "fmt"
import "math/rand"
import "time"
import "sort"

type Sequence []int

// Methods required by sort.Interface.
func (s Sequence) Len() int {
    return len(s)
}
func (s Sequence) Less(i, j int) bool {
    return s[i] < s[j]
}
func (s Sequence) Swap(i, j int) {
    s[i], s[j] = s[j], s[i]
}

// Bubble sort
func bubbleSort(arr Sequence) Sequence {
  newArr := make(Sequence, len(arr))
  copy(newArr, arr)
  arrLen := len(newArr)

  swapped := true
  for swapped {
    swapped = false
    for i := 0; i < arrLen - 1; i++ {
      if newArr[i + 1] < newArr[i] {
        temp := newArr[i]
        newArr[i] = newArr[i + 1]
        newArr[i + 1] = temp
        swapped = true
      }
    }
  }

  return newArr
}

// Create an array of a specified length with min and max range
func createArray(len, min, max int) Sequence {
  rand.Seed(time.Now().Unix())
  diff := max - min
  var arr Sequence = make(Sequence, len);
  for i := 0; i < len; i++ {
    arr[i] = rand.Int() % diff + min
  }
  return arr
}

// Create a sorted copy of an array using internal methods
func getSortedArray(arr Sequence) Sequence {
  newArr := make(Sequence, len(arr))
  copy(newArr, arr)
  sort.Sort(newArr)
  return newArr
}

// Compare 2 arrays
func compareArrays(arr1, arr2 Sequence) bool {
  arrLen := len(arr1)
  if arrLen != len(arr2) {
    fmt.Println("The array lengths do not match")
    return false
  }

  for i := 0; i < arrLen; i++ {
    if arr1[i] != arr2[i] {
      fmt.Println("The arrays did not match at index", i)
      return false
    }
  }

  return true
}

// Run the thing
func main() {
  arr := createArray(1000, 0, 10000)
  verificationArr := getSortedArray(arr)
  sortedArr := bubbleSort(arr)

  isSorted := compareArrays(verificationArr, sortedArr)
  if isSorted {
    fmt.Println("The sort worked!")
  } else {
    fmt.Println("THE SORT FAILED")
  }
}